nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py

```diff
@@ -430,3 +440,109 @@ def __iter__(self):
+    """
+    A dataset that accepts as argument multiple datasets and then samples from them based on the specified 
+    sampling technique.
+    Args:
+        datasets (list): A list of datasets to sample from.
+        shuffle (bool): Whether to shuffle individual datasets. Only works with non-iterable datasets. 
+            Defaults to True.
+        sampling_technique (str): Sampling technique to choose which dataset to draw a sample from.
+            Defaults to 'temperature'. Currently supports 'temperature' and 'round-robin'
+        sampling_temperature (int): Temperature value for sampling. Only used when sampling_technique = 'temperature'.
+            Defaults to 5.
+    """
+
+    def __init__(
+        self,
+        datasets: List[Any],
+        shuffle: bool = True,
+        sampling_technique: str = 'temperature',
+        sampling_temperature: int = 5,
+    ):
+        super().__init__()
+
+        supported_sampling_techniques = ['temperature', 'round-robin']
+        self.datasets = datasets
+        self.iterables = []
+        self.shuffle = shuffle
+        self.sampling_kwargs = {}
+        if sampling_technique == 'temperature':
+            self.index_generator = MultilingualTranslationDataset.temperature
+            self.sampling_kwargs['temperature'] = sampling_temperature
+        elif sampling_technique == 'round-robin':
+            self.index_generator = MultilingualTranslationDataset.round_robin
+        else:
+            raise ValueError(f"Currently we only support sampling techniques in {supported_sampling_techniques}.")
+        self.N = 0
+
+        if isinstance(datasets[0], IterableDataset):
+            self.kind = "iterable"
+        else:
+            self.kind = "map"
+
+        for dataset in datasets:
+            iterable = self.get_iterable(dataset)
+            self.iterables.append(iterable)
+            self.N += len(dataset)
+
+    def get_iterable(self, dataset):
+        if isinstance(dataset, IterableDataset):
+            return dataset.__iter__()
+        else:
+            if self.shuffle:
+                sampler = pt_data.RandomSampler(dataset)
+            else:
+                sampler = pt_data.SequentialSampler(dataset)
+            return sampler.__iter__()
+
+    def __iter__(self):
+        n = 0
+        ind_gen = self.index_generator(self.datasets, **self.sampling_kwargs)
+        while n < self.N:
+            n += 1
+            try:
+                ind = next(ind_gen)
+            except StopIteration:
+                return
+            try:
+                val = next(self.iterables[ind])
+                if self.kind == "map":
+                    val = self.datasets[ind][val]
+                yield val
+            except StopIteration:
+                self.iterables[ind] = self.get_iterable(self.datasets[ind])
+                if n < self.N:
+                    n -= 1
+
+    def __len__(self):
+        return self.N
+
+    @staticmethod
```

Comment:
Can we also add random sampling with specified sampling ratios irrespective of length and such?

Reply:
Yupp, sure! I can add a random sampler in the next commit. Though, we can keep adding more and more samplers as we keep on needing them. We can simply create them by following the same syntax as current samplers i.e. creating a static method that takes datasets list and kwargs as input.

Reply:
Added this in the update.

Reply:
Thanks for adding this :)